nexus/test-utils/src/lib.rs

```diff
@@ -139,247 +141,626 @@ pub async fn test_setup<N: NexusServer>(
     .await
 }
 
-pub async fn test_setup_with_config<N: NexusServer>(
-    test_name: &str,
-    config: &mut omicron_common::nexus_config::Config,
-    sim_mode: sim::SimMode,
-    initial_cert: Option<Certificate>,
-) -> ControlPlaneTestContext<N> {
-    let start_time = chrono::Utc::now();
-    let logctx = LogContext::new(test_name, &config.pkg.log);
-    let log = &logctx.log;
+struct RackInitRequestBuilder {
+    services: Vec<nexus_types::internal_api::params::ServicePutRequest>,
+    datasets: Vec<nexus_types::internal_api::params::DatasetCreateRequest>,
+    internal_dns_config: internal_dns::DnsConfigBuilder,
+}
+
+impl RackInitRequestBuilder {
+    fn new() -> Self {
+        Self {
+            services: vec![],
+            datasets: vec![],
+            internal_dns_config: internal_dns::DnsConfigBuilder::new(),
+        }
+    }
+
+    // Keeps track of:
+    // - The "ServicePutRequest" (for handoff to Nexus)
+    // - The internal DNS configuration for this service
+    fn add_service(
+        &mut self,
+        address: SocketAddrV6,
+        kind: ServiceKind,
+        service_name: internal_dns::ServiceName,
+        sled_id: Uuid,
+    ) {
+        let zone_id = Uuid::new_v4();
+        self.services.push(ServicePutRequest {
+            address,
+            kind,
+            service_id: Uuid::new_v4(),
+            sled_id,
+            zone_id: Some(zone_id),
+        });
+        let zone = self
+            .internal_dns_config
+            .host_zone(zone_id, *address.ip())
+            .expect("Failed to set up DNS for {kind}");
+        self.internal_dns_config
+            .service_backend_zone(service_name, &zone, address.port())
+            .expect("Failed to set up DNS for {kind}");
+    }
 
-    // Start up CockroachDB.
-    let database = db::test_setup_database(log).await;
+    // Keeps track of:
+    // - The "DatasetPutRequest" (for handoff to Nexus)
+    // - The internal DNS configuration for this service
+    fn add_dataset(
+        &mut self,
+        zpool_id: Uuid,
+        dataset_id: Uuid,
+        address: SocketAddrV6,
+        kind: DatasetKind,
+        service_name: internal_dns::ServiceName,
+    ) {
+        self.datasets.push(DatasetCreateRequest {
+            zpool_id,
+            dataset_id,
+            request: DatasetPutRequest { address, kind },
+        });
+        let zone = self
+            .internal_dns_config
+            .host_zone(dataset_id, *address.ip())
+            .expect("Failed to set up DNS for {kind}");
+        self.internal_dns_config
+            .service_backend_zone(service_name, &zone, address.port())
+            .expect("Failed to set up DNS for {kind}");
+    }
+}
+
+pub struct ControlPlaneTestContextBuilder<'a, N: NexusServer> {
```

Comment:
This cleanup is long overdue.  Thanks for taking it on!

The only real note I have here is that it's not that clear to me what use cases are supposed to be supported here and I'm afraid people it will become a footgun.  We obviously have the use case via `build()` that builds a full `ControlPlaneTestContext` out of what's here.  I gather the Builder itself exists to support more flexible constructions, like doing things in a different order or not starting some components.  Are you supposed to do this by using the `pub` fields of the Builder and never calling `build()`?  That seems a little confusing and error-prone.   But that may be practical if we really only expect this to be used in a few places!  Relatedly, it seems like you could misuse this in a lot of ways -- like calling `start_internal_nexus()` more than once.

I'm not really sure what to suggest instead.  Thinking out loud: hopefully, each of these components (DNS server, sled agent, nexus, etc.) already has a function that makes it easy to instantiate it in isolation.  The test setup is really providing the glue that (1) actually does it all for you in the common case; (2) handles the "glue", like DNS and rack init request setup, that is otherwise tedious and error-prone; and (3) makes sure that each thing is pointed at the copies of its dependencies that have been set up already.  That suggests maybe a `ControlPlaneBuilder` (which isn't testing-specific, really) with functions like `start_internal_dns_server()` and `start_nexus()` and `nexus_handoff()` -- probably very similar to what's here.  But it wouldn't have to use an explicit Builder that suggests (wrongly) that you're supposed to call `build()` or that you can't use the stuff you've instantiated without calling `build()`.  We could use typestates to ensure that you can't, say, do a handoff if you haven't started Nexus, and you also can't start more stuff after that.  On the other hand, using typestates this way might get in the way of some of the stuff we actually want to test.  So I'm less sure about that.

Take that as food for thought...the only real issue I have is the footgun fear, that people might mistake this for "the way to test anything other than what ControlPlaneTestContext gives you", and then winding up with spaghetti code dealing with a mess of `Option`s.  That's not a big deal if we really only expect this to be used in, like, the one place where you're using it in #3365.

Reply:
Given that this is test code, I significantly prefer to rely on assertions and calls to `.expect`, rather than typestates. My main reason for this disposition is: I struggle to predict all the interleavings someone might want to use when setting up a test.

Maybe we want to spin up Nexus without a sled agent? Maybe we want to spin up Nexus with ten sled agents? Maybe we want to spin up *almost* everything, but take a particular dropshot server offline right before "nexus external" is created?

Regardless of what we're trying to test, my goals were:
- de-duplicate the (massive!) amount of code necessary for setup
- make it easy to do the right thing
- make it *possible* to test weirder things

That's part of why I used so many `Option` types - the builder lets the caller "construct whatever state they want", and, if really desired, access those types via `pub` fields.

I agree with you that this is an API which "could let a caller do the wrong thing", but I also struggle with that, because we also **want** to test weird behaviors during tests, whenever possible.

If I'm overlooking an alternative, I'm totally open to suggestions! But providing a "single-use, hard-to-get-wrong API for test setup" that sacrifices flexibility is basically what we had before.